relation/src/gadgets/ecc/emulated/short_weierstrass.rs

```diff
@@ -0,0 +1,422 @@
+
+        // Case 2: p2 is infinity, while p0 and p1 are not.
+        // inf0 == 1 || inf1 == 1 || inf2 == 0 || x0 != x1 || (y0 == y1 && y0 != 0)
+        let non_inf_p2 = self.logic_neg(p2.2)?;
+        let eq_x0_x1 = self.is_emulated_var_equal(&p0.0, &p1.0)?;
+        let neq_x0_x1 = self.logic_neg(eq_x0_x1)?;
+        let eq_y0_y1 = self.is_emulated_var_equal(&p0.1, &p1.1)?;
+        let is_y0_zero = self.is_emulated_var_zero(&p0.1)?;
+        let not_y0_zero = self.logic_neg(is_y0_zero)?;
+        let t = self.logic_and(eq_y0_y1, not_y0_zero)?;
+        let t = self.logic_or(neq_x0_x1, t)?;
+        let t = self.logic_or(non_inf_p2, t)?;
+        self.logic_or_gate(infinity_mark, t)?;
+
+        // Case 3: point doubling
+        // doubling mark is 1 iff x0 == x1 and y0 == y1
+        let doubling_mark = self.mul(eq_x0_x1.0, eq_y0_y1.0)?;
+        let doubling_coef = self.mul(doubling_mark, non_infinity_mark.0)?;
+        let doubling_coef = self.mul(doubling_coef, not_y0_zero.0)?;
+        // forcefully convert Variable into EmulatedVariable
+        // safe because it's boolean
+        let mut v = vec![self.zero(); E::NUM_LIMBS];
+        v[0] = doubling_coef;
+        let doubling_coef = EmulatedVariable::<E>(v, core::marker::PhantomData);
+
+        // first equality (x2 + 2 * x0) * (y0 + y0)^2 == (3 * x1^2 + a)^2
+        let y0_times_2 = self.emulated_add(&p0.1, &p0.1)?;
+        let x0_plus_x1 = self.emulated_add(&p0.0, &p1.0)?;
+        let x0_plus_x1_plus_x2 = self.emulated_add(&p2.0, &x0_plus_x1)?;
+        let lhs = self.emulated_mul(&x0_plus_x1_plus_x2, &y0_times_2)?;
+        let lhs = self.emulated_mul(&lhs, &y0_times_2)?;
+        // s = 3 * x1^2 + a
+        let s = self.emulated_mul(&p0.0, &p0.0)?;
+        let s = self.emulated_mul_constant(&s, E::from(3u64))?;
+        let s = self.emulated_add_constant(&s, a)?;
+        let rhs = self.emulated_mul(&s, &s)?;
+
+        let lhs = self.emulated_mul(&lhs, &doubling_coef)?;
+        let rhs = self.emulated_mul(&rhs, &doubling_coef)?;
+        self.enforce_emulated_var_equal(&lhs, &rhs)?;
+
+        // second equality (y2 + y0) * (y0 + y0) == (3 * x1^2 + a) (x0 - x2)
+        let y2_plus_y0 = self.emulated_add(&p2.1, &p0.1)?;
+        let lhs = self.emulated_mul(&y2_plus_y0, &y0_times_2)?;
+        let x0_minus_x2 = self.emulated_sub(&p0.0, &p2.0)?;
+        let rhs = self.emulated_mul(&s, &x0_minus_x2)?;
+
+        let lhs = self.emulated_mul(&lhs, &doubling_coef)?;
+        let rhs = self.emulated_mul(&rhs, &doubling_coef)?;
+        self.enforce_emulated_var_equal(&lhs, &rhs)?;
+
+        // Case 4: point addition
+        let coef = self.mul(non_infinity_mark.0, neq_x0_x1.0)?;
+        // forcefully convert Variable into EmulatedVariable
+        // safe because it's boolean
+        let mut v = vec![self.zero(); E::NUM_LIMBS];
+        v[0] = coef;
+        let coef = EmulatedVariable::<E>(v, core::marker::PhantomData);
```

Comment:
maybe call it `add_coef`

Reply:
Done in [cf9ff0e](https://github.com/EspressoSystems/jellyfish/pull/326/commits/cf9ff0ee56440e4767b8738978ab82b34c7afc32)